Makefile

```diff
@@ -120,181 +135,141 @@ build/src/libultra/libc/ll%: OPTFLAGS := -O1
+#### Main Targets ###
 
-# just using build/baserom still probably has some race condiction/dependency bug, but since
-# it is first and should be completed relatively fast, it should not occur all that often.
-$(UNCOMPRESSED_ROM): build/baserom $(TEXTURE_FILES_OUT) $(ASSET_FILES_OUT) $(UNCOMPRESSED_ROM_FILES)
-	./tools/makerom.py ./tables/dmadata_table.txt $@
+uncompressed: $(ROM)
 ifeq ($(COMPARE),1)
-	@md5sum $(UNCOMPRESSED_ROM)
+	@md5sum $(ROM)
 	@md5sum -c checksum_uncompressed.md5
 endif
 
-$(ROM): build/baserom $(ROM_FILES)
-	./tools/makerom.py ./tables/dmadata_table.txt $@ -c
+compressed: $(ROMC)
 ifeq ($(COMPARE),1)
-	@md5sum $(ROM)
+	@md5sum $(ROMC)
 	@md5sum -c checksum.md5
 endif
 
-all: $(UNCOMPRESSED_ROM) $(ROM) ;
-
-build/code.elf: $(O_FILES) build/linker_scripts/code_script.ld undef.txt build/linker_scripts/object_script.ld build/dmadata_script.ld
-	$(LD) -T build/linker_scripts/code_script.ld -T undef.txt -T build/linker_scripts/object_script.ld -T build/dmadata_script.ld --no-check-sections --accept-unknown-input-arch -Map build/mm.map -N -o $@
-
-build/code_pre_dmadata.elf: $(O_FILES) build/linker_scripts/code_script.ld undef.txt build/linker_scripts/object_script.ld
-	$(LD) -r -T build/linker_scripts/code_script.ld -T undef.txt -T build/linker_scripts/object_script.ld --no-check-sections --accept-unknown-input-arch -N -o $@
-
-build/dmadata_script.txt: tables/dmadata_table.txt build/code_pre_dmadata.elf
-	./tools/dmadata.py ./tables/dmadata_table.txt /dev/null -u -l $@ -e build/code_pre_dmadata.elf
-
-build/dmadata: $(ROM_FILES:build/dmadata=)
-	./tools/dmadata.py ./tables/dmadata_table.txt $@
-
-build/uncompressed_dmadata: $(UNCOMPRESSED_ROM_FILES:build/uncompressed_dmadata=)
-	./tools/dmadata.py ./tables/dmadata_table.txt $@ -u
-
-build/binary/boot build/binary/code: build/code.elf
-	$(OBJCOPY) --dump-section $(notdir $@)=$@ $< /dev/null
-
-build/binary/assets/scenes/%: build/code.elf
-	$(OBJCOPY) --dump-section $*=$@ $< /dev/null
-
-build/binary/overlays/%: build/code.elf
-	$(OBJCOPY) --dump-section $*=$@ $< /dev/null
+.PHONY: all uncompressed compressed clean assetclean distclean disasm init setup
 
+# make will delete any generated assembly files that are not a prerequisite for anything, so keep it from doing so
+.PRECIOUS: asm/%.asm
+.DEFAULT_GOAL := uncompressed
 
-#### ASM rules ####
+all: uncompressed compressed
 
-# Use an empty sentinel file (dep) to track the directory as a dependency, and
-# emulate GNU Make's order-only dependency.
-# The `touch $@; action || rm $@` pattern ensures that the `dep` file is older
-# than the output files from `action`, and only exists if `action` succeeds.
-asm/non_matchings/%/dep: asm/%.asm
-	@mkdir -p $(dir $@)
-	@touch $@
-	./tools/split_asm.py $< asm/non_matchings/$* || rm $@
+$(ROM): $(ELF)
+	$(ELF2ROM) -cic 6105 $< $@
 
-asm/%.asm: asm/disasm.dep ;
+$(ROMC): $(ROM)
+	python3 tools/z64compress_wrapper.py --mb 32 --matching --threads $(shell nproc) $< $@ $(ELF) build/$(SPEC)
```

Comment:
So I'm not a huge fan of always using max available threads for compression. I feel this should be left as an argument for the user to specify. I think best option (though do not know if possible) would be parse if the -j option was passed to make. If not don't use multithreading, if yes without a number specified use max possible threads, and if yes with number just pass that number along.

Reply:
Just clarifying that while I believe this should be changed, this should not prevent the NBS from being merged as it currently stands, especially if it would take significant time and effort to "fix" this.

Reply:
From what I remember, getting the value `-j` passed to `make` was non-trivial :s

Reply:
Also, I may be misremembering but I think `nproc` doesn't work in Mac, right?
Can somebody check?

Reply:
I remember we had a long conversation about this in Discord relating to passing `j` to making tools, and I ended up writing some very messy thing that should work on most architectures. Not worth it IMO, may as well specify a fixed number of threads, with a make arg to change it. This is the PR I made that started it: https://github.com/zeldaret/oot/pull/804 , and this is where the conversation started: https://discord.com/channels/688807550715560050/688851317593997489/838802089017868369

Reply:
I agree this would be the second best option, though the default should be no multithreading imo.

Reply:
It's worthy to note that `extract_assets.py` extracts with as many cores as it can as well. If this is to be changed it stands to reason that we'd have to address that case as well.